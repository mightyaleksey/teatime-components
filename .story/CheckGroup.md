## Preview


### size m

```
<div>
  some text for the baseline
  <CheckGroup
    cols={3}
    defaultValue={[
      'kawasaki',
      'vespa'
    ]}
    name='checkGroup-m'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='m'/>
  {' '}
  <CheckGroup
    cols={3}
    disabled={true}
    name='checkGroup-m-d'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='m'/>
</div>
```


### size l

```
<div>
  some text for the baseline
  <CheckGroup
    cols={3}
    name='checkGroup-l'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='l'/>
  {' '}
  <CheckGroup
    cols={3}
    disabled={true}
    name='checkGroup-l-d'
    options={[
      {label: 'Ducati', value: 'ducati'},
      {label: 'Honda', value: 'honda'},
      {label: 'Jawa', value: 'jawa'},
      {label: 'Kawasaki', value: 'kawasaki'},
      {label: 'Suzuki', value: 'suzuki'},
      {label: 'Ural', value: 'ural'},
      {label: 'Vespa', value: 'vespa'},
      {label: 'Yamaha', value: 'yamaha'}
    ]}
    size='l'/>
</div>
```
