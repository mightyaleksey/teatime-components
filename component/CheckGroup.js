'use strict';

const {Component, PropTypes} = require('react');
const {assign, map, noop} = require('../lib/dash');
const {filterProps, genericName, isControlled} = require('../lib/util');
const Box = require('../view/Box');
const React = require('react');
const cc = require('classnames');
const chunk = require('lodash.chunk');

const cssModules = {
  l: require('../style/checkGroup/checkGroup-l.css'),
  m: require('../style/checkGroup/checkGroup-m.css'),
};

class CheckGroup extends Component {
  constructor(props) {
    super(props);

    this._controlled = isControlled(props);

    const checkItems = this._checkItems = calculateCheckItems(props.options);
    const value = this.controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      values: mapValueToState(checkItems, value || []),
    };
  }

  componentWillReceiveProps(nextProps) {
    this._controlled = isControlled(nextProps);

    const {options} = this.props;
    const value = this._controlled
      ? nextProps.value
      : this.state.value;

    var values = this.state.values;

    if (nextProps.options !== options) {
      this._checkItems = calculateCheckItems(nextProps.options);
      values = mapValueToState(this._checkItems, value || []);
    }

    this.setState({values});
  }

  css = tokenName => genericName(this.props, tokenName)

  focus = noop

  _onChange = (e, {checked}, position) => {
    const values = updateValue(this.state.values, position, checked);

    if (!this.controlled) {
      this.setState({values});
    }

    this.props.onChange(e, {
      value: mapStateToValue(this._checkItems, values),
    });
  }

  computeChecks(checkItems) {
    const {disabled, name} = this.props;
    const {values} = this.state;

    const styles = this.css();

    return map(item =>
      this.renderCheck({
        ...item,
        checked: values[item._position],
        disabled,
        key: `${item._position}${item.value}`,
        name,
        onChange: this._onChange,
        position: item._position,
        styles,
      }), checkItems);
  }

  computeColumns(checks) {
    const {cols} = this.props;
    const chunkSize = Math.ceil(this._checkItems.length / cols);
    const columns = chunk(checks, chunkSize);
    const className = this.css('column');

    var index = 0;
    return map(children =>
      this.renderColumn({
        children,
        className,
        key: ++index,
      }), columns);
  }

  render() {
    const {className} = this.props;
    const {css} = this;

    const checks = this.computeChecks(this._checkItems);
    const columns = this.computeColumns(checks);

    return (
      <div
        {...filterProps(this.props)}
        className={cc(css('container'), className)}
        onChange={void 0}>
        {columns}
      </div>
    );
  }

  renderColumn(columnProps) {
    return (
      <div {...columnProps}/>
    );
  }

  renderCheck(checkProps) {
    return (
      <Box {...checkProps}/>
    );
  }
}

CheckGroup.defaultProps = {
  cols: 1,
  onChange: noop,
  size: 'm',
  styles: cssModules,
};

CheckGroup.propTypes = {
  cols: PropTypes.number,
  defaultValue: PropTypes.arrayOf(PropTypes.string),
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  options: PropTypes.array.isRequired,
  size: PropTypes.oneOf([
    'l',
    'm',
  ]),
  styles: PropTypes.object,
  value: PropTypes.arrayOf(PropTypes.string),
};

module.exports = CheckGroup;

// calculateMenuItems :: (a -> q -> bool) -> [a] -> q -> [b]
function calculateCheckItems(items = []) {
  const length = items.length;
  const nextItems = Array(length);

  for (var i = 0; i < length; ++i)
    nextItems[i] = assign(items[i], {_position: i});

  return nextItems;
}

function mapStateToValue(options, values) {
  const selected = [];

  for (var i = 0; i < values.length; ++i) {
    if (!values[i]) {
      continue;
    }

    selected.push(options[i].value);
  }

  return selected;
}

function mapValueToState(options, selected) {
  const selectedMap = {};

  var length = selected.length;
  while (length--) {
    selectedMap[selected[length]] = null;
  }

  return map(({value}) => value in selectedMap, options);
}

function updateValue(values, position, target) {
  return values.map((value, i) => position !== i ? value : target);
}
