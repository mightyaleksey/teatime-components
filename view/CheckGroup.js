'use strict';

const { Component, PropTypes } = require('react');
const { bind, hasValueProp, sanitizeProps } = require('../tool/component');
const { generateId, hasUniqueValues, mapKey, mapKeyBasedOnPos } = require('../tool/identity');
const { isUndefined, mapRange, noop } = require('../tool/func');
const { styleName } = require('../tool/className');
const Check = require('./Check');
const React = require('react');
const warning = require('../tool/warning');

var didWarnForDefaultValue = false;

class CheckGroup extends Component {
  constructor(props) {
    super(props);

    bind(this, 'onChange');

    this.controlled = hasValueProp(props);

    if (process.env.NODE_ENV !== 'production' && this.controlled && !didWarnForDefaultValue) { // eslint-disable-line no-undef
      warning(isUndefined(props.defaultValue),
        'CheckGroup elements must be either controlled or uncontrolled ' +
        '(specify either the value prop, or the defaultValue prop, but not ' +
        'both). Decide between using a controlled or uncontrolled select ' +
        'element and remove one of these props. More info: ' +
        'https://fb.me/react-controlled-components');

      didWarnForDefaultValue = true;
    }

    this.updateKeyMapper(props.hasUniqValues, props.options);

    const value = this.controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      prefix: generateId(),
      values: mapValueToState(props.options, value || []),
    };
  }

  componentWillReceiveProps({ hasUniqValues, options, value }) {
    if (this.controlled) {
      this.setState({values: mapValueToState(options, value)});
    }

    if (this.props.hasUniqValues !== hasUniqValues) {
      this.updateKeyMapper(hasUniqValues, options);
    }
  }

  onChange(e, { checked }, tc) {
    const values = updateValue(this.state.values, tc, checked);

    if (!this.controlled) {
      this.setState({values});
    }

    this.props.onChange(e, {
      value: mapStateToValue(this.props.options, values),
    });
  }

  /**
   * @param {boolean} hasUniqValues
   * @param {object[]} options
   */
  updateKeyMapper(hasUniqValues, options) {
    this.mapKey = !(hasUniqValues && hasUniqueValues(options))
      ? mapKeyBasedOnPos
      : mapKey;
  }

  render() {
    return (
      <div
        {...sanitizeProps(['options', 'styles', 'hasUniqValues', 'styleName'], this.props)}
        className={styleName(this.props)}
        onChange={undefined}>
        {this.renderColumns()}
      </div>
    );
  }

  renderColumns() {
    const { cols, styles } = this.props;
    const { prefix } = this.state;
    const rCols = Math.max(cols || 0, 1);

    if (rCols === 1) {
      return this.renderOptions(0, rCols);
    }

    return mapRange(step => (
      <div
        className={styles.column}
        key={mapKeyBasedOnPos(prefix, '_', step)}>
        {this.renderOptions(step, rCols)}
      </div>
    ), rCols);
  }

  /**
   * @param  {number} start
   * @param  {number} step
   * @return {component[]}
   */
  renderOptions(start, step) {
    const { disabled: globalDisabled, name, options, styles } = this.props;
    const { prefix, values } = this.state;

    const result = [];

    for (var i = start; i < options.length; i += step) {
      result.push((
        <Check
          {...options[i]}
          disabled={globalDisabled || options[i].disabled}
          checked={values[i]}
          key={this.mapKey(prefix, options[i].value, i)}
          name={name}
          onChange={this.onChange}
          styles={styles}
          tc={i}/>
      ));
    }

    return result;
  }
}

CheckGroup.defaultProps = {
  hasUniqValues: true,
  onChange: noop,
  styleName: 'container',
  styles: {},
};

CheckGroup.propTypes = {
  cols: PropTypes.number,
  defaultValue: PropTypes.arrayOf(PropTypes.string),
  hasUniqValues: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onContextMenu: PropTypes.func,
  options: PropTypes.array.isRequired,
  styleName: PropTypes.string,
  styles: PropTypes.shape({
    column: PropTypes.string,
    container: PropTypes.string,
    control: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    native: PropTypes.string.isRequired,
    wrapper: PropTypes.string.isRequired,
  }),
  value: PropTypes.arrayOf(PropTypes.string),
};

module.exports = CheckGroup;

/**
 * @param  {object[]} options
 * @param  {boolean[]} values
 * @return {string[]}
 */
function mapStateToValue(options, values) {
  const selected = [];

  for (var i = 0; i < values.length; ++i) {
    if (!values[i]) {
      continue;
    }

    selected.push(options[i].value);
  }

  return selected;
}

/**
 * @param  {object[]} options
 * @param  {string[]} selected
 * @return {boolean[]}
 */
function mapValueToState(options, selected) {
  const selectedMap = {};

  var length = selected.length;
  while (length--) {
    selectedMap[selected[length]] = null;
  }

  return options.map(({ value }) => value in selectedMap);
}

/**
 * @param  {boolean[]} values
 * @param  {number} position
 * @param  {boolean} target
 * @return {boolean[]}
 */
function updateValue(values, position, target) {
  return values.map((value, i) => position !== i ? value : target);
}
