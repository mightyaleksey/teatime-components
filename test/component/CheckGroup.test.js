import CheckGroup from '../../component/CheckGroup';
import React from 'react';
import renderer from 'react-test-renderer';

test('CheckGroup renders correctly with defaults', () => {
  const tree = renderer.create(
    <CheckGroup
      name='checkGroup control'
      options={[
        {label: 'Ducati', value: 'ducati'},
        {label: 'Honda', value: 'honda'},
        {label: 'Jawa', value: 'jawa'},
      ]}/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('CheckGroup renders correctly with defined props', () => {
  const tree = renderer.create(
    <CheckGroup
      cols={2}
      id='checkGroups id'
      name='checkGroup control'
      options={[
        {label: 'Ducati', value: 'ducati'},
        {disabled: true, label: 'Honda', value: 'honda'},
        {label: 'Jawa', value: 'jawa'},
      ]}
      value={[
        'ducati',
        'honda',
      ]}/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
