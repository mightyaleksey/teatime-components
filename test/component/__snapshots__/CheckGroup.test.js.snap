exports[`test CheckGroup renders correctly with defaults 1`] = `
<div
  className="container"
  id={undefined}
  onChange={undefined}>
  <div
    className="column">
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={undefined}
        id="_teatime1"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="ducati" />
      <label
        className="control"
        htmlFor="_teatime1" />
      <label
        className="label"
        htmlFor="_teatime1">
        Ducati
      </label>
    </span>
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={undefined}
        id="_teatime2"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="honda" />
      <label
        className="control"
        htmlFor="_teatime2" />
      <label
        className="label"
        htmlFor="_teatime2">
        Honda
      </label>
    </span>
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={undefined}
        id="_teatime3"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="jawa" />
      <label
        className="control"
        htmlFor="_teatime3" />
      <label
        className="label"
        htmlFor="_teatime3">
        Jawa
      </label>
    </span>
  </div>
</div>
`;

exports[`test CheckGroup renders correctly with defined props 1`] = `
<div
  className="container"
  id="checkGroups id"
  onChange={undefined}>
  <div
    className="column">
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={undefined}
        id="_teatime4"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="ducati" />
      <label
        className="control"
        htmlFor="_teatime4" />
      <label
        className="label"
        htmlFor="_teatime4">
        Ducati
      </label>
    </span>
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={true}
        id="_teatime5"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="honda" />
      <label
        className="control"
        htmlFor="_teatime5" />
      <label
        className="label"
        htmlFor="_teatime5">
        Honda
      </label>
    </span>
  </div>
  <div
    className="column">
    <span
      className="wrapper">
      <input
        checked={false}
        className="native"
        defaultValue={undefined}
        disabled={undefined}
        id="_teatime6"
        name="checkGroup control"
        onChange={[Function]}
        type="checkbox"
        value="jawa" />
      <label
        className="control"
        htmlFor="_teatime6" />
      <label
        className="label"
        htmlFor="_teatime6">
        Jawa
      </label>
    </span>
  </div>
</div>
`;
